### PR TITLE
feat: send fee tokens flow

### PR DESCRIPTION
**Pull request by raul-oliveira**
This is the implementation of the send tokens with fee based tokens interaction.

### Acceptance Criteria
  - Fee calculation: When sending fee-based tokens, the network fee (in HTR) is calculated in real-time as the user fills outputs;
  deposit-based tokens and HTR show a "No fee" badge. Send button is disabled if insufficient HTR for fees.
  - Fee recalculation: Fee updates automatically when outputs are added/removed/edited, token selection changes, or input mode switches.
  - Transaction Overview modal: Before confirming (software wallet only), a modal shows all outputs (with change marked), network fee, total payment summary, and requires a 6-digit PIN to confirm.
  - Send success modal: After successful send, shows sent amounts per token with options to close (resets form) or view transaction  details.
  - Error handling: If transaction fails, the overview modal shows the error with "Try again" (resets PIN) and "Cancel" options. 

### Security Checklist
- [x] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

image
image
image
image
image

## Summary by CodeRabbit

* **New Features**
  * Transaction overview modal with fee breakdown, 6-digit PIN confirmation, multi-phase review/sending/error flow
  * Send success modal summarizing tokens sent and offering view-details action

* **UI**
  * Per-token fee handling (auto/manual) with inline fee calculation, “You’ll pay” summary and insufficiency errors that can block sending
  * Streamlined outputs add/remove controls and live input handling
  * Transaction details now show fee paid per token and fee presentation/badges/messages

* **Localization**
  * Many new and updated user-facing strings across send/fee/transaction flows

* **Tests**
  * Improved E2E selectors for backup/word-selection tests

### CODE REVIEW

File: src/components/ModalTransactionOverview.js [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    return Array.from(groups.values());
+  }, [outputs, tokens]);
+
+  /**
+   * Check if any token in the transaction requires a fee
+   */
+  const hasAnyFeeToken = useMemo(() => {
+    return groupedOutputs.some((group) => group.token.version === TOKEN_VERSION_FEE);
```
Comment (raul-oliveira) [inline src/components/ModalTransactionOverview.js]: use the hathorlib constant

File: src/components/SendTokensOne.js [hunk 2]
```diff
@@ -248,6 +483,80 @@ class SendTokensOne extends React.Component {
...
+                      fontSize: '10px',
+                      padding: '4px 8px',
+                      borderRadius: '4px',
+                      whiteSpace: 'nowrap',
+                      zIndex: 1000,
+                      lineHeight: '16px',
+                    }}
+                  >
+                    {hathorLib.tokensUtils.isHathorToken(this.state.selected.uid)
+                      ? t`This is the native token, no network fee will be charged.`
```
Comment (raul-oliveira) [inline src/components/SendTokensOne.js]: remove this if ternary. it should be in a component or function

File: src/screens/SendTokens.js [hunk 3]
```diff
@@ -359,36 +400,82 @@ function SendTokens() {
...
-              pin,
-              prepareSendTransaction: prepareSendTransaction,
-              onSendSuccess: onSendSuccess,
-              onSendError: onSendError,
-              title: t`Sending transaction`,
-            });
-          }
-        })
+        // Calculate total fee
+        const totalFee = Object.values(tokenFees).reduce((a, b) => a + b, 0n);
```
Comment (raul-oliveira) [inline src/screens/SendTokens.js]: to open the modal, we should call prepareTx without signing the tx. when the user confirm we sign and send.

File: src/screens/SendTokens.js [hunk 4]
```diff
@@ -439,6 +526,58 @@ function SendTokens() {
...
+      value: output.value,
+      token: output.token,
+    }));
+  };
+
+  /**
+   * Get total amount for a token from outputs
+   */
+  const getTotalAmountForToken = (tokenUid, formOutputs) => {
+    return formOutputs
```
Comment (raul-oliveira) [inline src/screens/SendTokens.js]: is possible to store the outputs in a map (dict) to have it already grouped by token?

File: src/components/SendTokensOne.js [hunk 5]
```diff
@@ -167,14 +396,59 @@ class SendTokensOne extends React.Component {
...
+        </span>
+      );
+    }
+
+    return (
+      <span>
+        {hathorLib.numberUtils.prettyValue(this.state.fee, this.props.decimalPlaces)} HTR
+      </span>
+    );
+  }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/SendTokensOne.js:447]: Done

File: src/components/TxData.js [hunk 6]
```diff
@@ -289,6 +291,35 @@ class TxData extends React.Component {
...
+      const parsedTx = hathorLib.Transaction.createFromBytes(txBytes, network);
+      const feeHeader = parsedTx.getFeeHeader();
+      if (!feeHeader) return;
+
+      const feeEntries = feeHeader.entries.map(entry => {
+        const tokenConfig = this.getOutputToken(entry.tokenIndex);
+        return {
+          tokenSymbol: tokenConfig?.symbol || 'Unknown',
+          amount: entry.amount,
+        };
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/TxData.js:313]: Done

File: src/hooks/useTokenDetails.js [hunk 7]
```diff
@@ -11,27 +11,38 @@ import { tokenRegisterRequested } from '../actions';
...
+    if (reg?.status === TOKEN_DOWNLOAD_STATUS.FAILED) {
+      errors[uid] = reg.error;
+    }
+  }
+
+  const uidsNeedingFetch = tokens
+    .filter(t => t.version === undefined && !errors[t.uid])
+    .map(t => t.uid);
+
+  const isLoading = uidsNeedingFetch.length > 0;
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/hooks/useTokenDetails.js]: Done

File: src/components/ModalTransactionOverview.js [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+      } else {
+        tokenTotals.set(htrUid, { symbol: 'HTR', total: fee });
+      }
+    }
+
+    const parts = [];
+    for (const [, { symbol, total }] of tokenTotals) {
+      parts.push(`${hathorLib.numberUtils.prettyValue(total, decimalPlaces)} ${symbol}`);
+    }
+    return parts.join(' + ');
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/ModalTransactionOverview.js:87]: won't do.

File: src/components/TxData.js [hunk 9]
```diff
@@ -289,6 +291,35 @@ class TxData extends React.Component {
...
+   */
+  parseFeeHeader = () => {
+    try {
+      const { raw } = this.props.transaction;
+      if (!raw) return;
+
+      const network = hathorLib.config.getNetwork();
+      const txBytes = Buffer.from(raw, 'hex');
+      const parsedTx = hathorLib.Transaction.createFromBytes(txBytes, network);
+      const feeHeader = parsedTx.getFeeHeader();
```
Comment (pedroferreira1) [inline src/components/TxData.js:305]: If I'm not wrong, the API to fetch tx detail also returns the headers, doesn't it?

File: src/hooks/useTokenDetails.js [hunk 10]
```diff
@@ -11,27 +11,38 @@ import { tokenRegisterRequested } from '../actions';
...
+ * Hook to fetch token details (including version) on-demand for multiple tokens.
+ * Automatically dispatches tokenRegisterRequested for each token whose version is undefined.
  *
- * @param {string|undefined} uid - Token uid to fetch details for
- * @returns {{ token: object|undefined, isLoading: boolean, error: string|null }}
+ * @param {string[]} uids - Array of token uids to fetch details for
+ * @returns {{ tokens: Array<{uid: string, name: string, symbol: string, version: number|undefined}>, isLoading: boolean, errors: Object<string, string|null> }}
  */
-export function useTokenDetails(uid) {
+export function useTokensDetails(uids) {
```
Comment (pedroferreira1) [inline src/hooks/useTokenDetails.js:20]: What's this multi token hook doing in the send tokens flow? Or is it fixing another bug? To be honest, I think the code became a bit too complex, that's why I want to understand why do you need this change, to think if we could do in another way

File: src/components/OutputsWrapper.js [hunk 11]
```diff
@@ -73,7 +93,9 @@ class OutputsWrapper extends React.Component {
         <input type="datetime-local" placeholder={t`Date and time in GMT`} step="1"
           className="form-control output-timelock col-3" style={{display: 'none'}} ref={this.timelock}
           disabled={LOCAL_STORE.isHardwareWallet() ? true : null}/>
-        {this.props.index === 0 ? <button type="button" className="btn btn-hathor" onClick={this.props.addOutput}>+</button> : null}
+        {this.props.index === 0
+          ? <button type="button" className="btn btn-hathor" onClick={this.props.addOutput}>+</button>
+          : <button type="button" className="btn btn-hathor" onClick={() => this.props.removeOutput(this.props.index)}>-</button>}
```
Comment (pedroferreira1) [inline src/components/OutputsWrapper.js]: I know that this was not on figma but I'm not sure I like this UI. It's too similar compared to the + button. I would make a bit different, maybe an X without the whole btn UI we have. What do you think?

File: src/screens/SendTokens.js [hunk 12]
```diff
@@ -45,13 +46,14 @@ function SendTokens() {
   const navigate = useNavigate();
 
   // Redux state
-  const { selectedToken, tokens, metadataLoaded, useWalletService } = useSelector(
+  const { selectedToken, tokens, useWalletService, tokensBalance, decimalPlaces } = useSelector(
     (state) => {
       return {
         selectedToken: state.selectedToken,
         tokens: state.tokens,
-        metadataLoaded: state.metadataLoaded,
```
Comment (pedroferreira1) [inline src/screens/SendTokens.js:53]: Why did you remove this? It's important for us to know the token metadata, so we can know the ones that are NFTs. How are you doing this validation without this metadata?

File: src/components/TxData.js (hunk 9 above)
Comment (raul-oliveira) [reply src/components/TxData.js:305]: Not yet. Here is the pr that address it https://github.com/HathorNetwork/hathor-core/pull/1646

File: src/hooks/useTokenDetails.js (hunk 10 above)
Comment (raul-oliveira) [reply src/hooks/useTokenDetails.js:20]: It's actually a guard. When the user doesn't have updated tokens (with the version field) in the wallet, we call the register token saga. This saga will fetch from memory store first, and as a fall back will try the `getTokenDetails` api.

The change on this guy, is that this hook was used to get exactly one token before, now we need more than one.

File: src/components/OutputsWrapper.js (hunk 11 above)
Comment (raul-oliveira) [reply src/components/OutputsWrapper.js]: I'll test something and send you a screenshot for us to discuss.

File: cypress/e2e/00-welcome.cy.js
Comment (tuliomir) [inline cypress/e2e/00-welcome.cy.js:1]: question: Always nice to have improvements on the test suites, but is this related to this PR in any way?

Probably best to do this in another PR

File: src/components/ModalSendSuccess.js [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+              {t`Transaction sent`}
+            </h5>
+            <button type="button" className="close" onClick={handleClose} aria-label="Close">
+              <span aria-hidden="true">&times;</span>
+            </button>
+          </div>
+          <div className="modal-body">
+            <p>{t`Your transaction was sent successfully!`}</p>
+            <p>
+              {t`You successfully sent`} <strong>{formatSentAmounts()}</strong>.
```
Comment (tuliomir) [inline src/components/ModalSendSuccess.js:82]: question(non-blocking): How would this screen render sending 255 outputs of different tokens, all with amount over `Number.MAX_SAFE_INTEGER`?

It would be nice to have a test for those edge cases

File: src/hooks/useTokenDetails.js [hunk 14]
```diff
@@ -11,27 +11,38 @@ import { tokenRegisterRequested } from '../actions';
...
+
+  const uidsNeedingFetch = tokens
+    .filter(t => t.version === undefined && !errors[t.uid] && tokenRegistration[t.uid]?.status !== TOKEN_DOWNLOAD_STATUS.LOADING)
+    .map(t => t.uid);
+
+  const isLoading = uidsNeedingFetch.length > 0;
 
   useEffect(() => {
-    if (needsVersionFetch) {
+    uidsNeedingFetch.forEach(uid => {
```
Comment (tuliomir) [inline src/hooks/useTokenDetails.js]: question(non-blocking): Are those requests idempotent? I mean, if I re-request a token that was requested before, is it a no-op?

I can imagine that with that code, for every `uid` that no longer needs fetching the entire array is rebuilt and a new bulk of re-requests happen. So for `n` tokens we can get up to `n(n+1)/2` requests, which is quite a lot.

If my understanding of this code is correct, this would benefit from a simplification refactor.

File: src/components/ModalTransactionOverview.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (tuliomir) [reply src/components/ModalTransactionOverview.js:87]: question: Why not? 👀  Maybe this will become important once NFT tests are added here.

File: src/components/ModalTransactionOverview.js [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+  const renderReviewBody = () => (
+    <>
+      {/* Flat output list: regular outputs, then change, then data */}
+      <div className="mb-3">
+        {[
+          ...outputs.filter(o => o.type !== 'data' && !o.isChangeOutput),
+          ...outputs.filter(o => o.type !== 'data' && o.isChangeOutput),
+          ...outputs.filter(o => o.type === 'data'),
+        ].map((output, idx) => renderOutput(output, idx))}
+      </div>
```
Comment (tuliomir) [inline src/components/ModalTransactionOverview.js:272]: question(non-blocking): Same question here: how would this render with 255 tokens and astronomical amounts?

File: src/components/SendTokensOne.js [hunk 16]
```diff
@@ -167,14 +396,59 @@ class SendTokensOne extends React.Component {
...
+      this.calculateFeeAndChange();
+    });
     this.props.tokenSelectChange(selected, this.props.index);
   }
 
+  /**
+   * Returns the tooltip text for the fee info icon based on the selected token type
+   */
+  getFeeTooltipText = () => {
+    if (hathorLib.tokensUtils.isHathorToken(this.state.selected.uid)) {
```
Comment (tuliomir) [inline src/components/SendTokensOne.js]: suggestion(non-blocking): Create a helper state to make it easier to get the correct info. Future maintenance agents may pick the wrong set of properties to try to make this assumption in the future
```js
state: {
  ...
  isNativeToken: hathorLib.tokensUtils.isHathorToken(this.state.selected?.uid)
}  
...

    if (this.state.isNativeToken) {
```

File: src/hooks/useTokenDetails.js (hunk 14 above)
Comment (raul-oliveira) [reply src/hooks/useTokenDetails.js]: Yes they are because we are fetching first from memory store then by the api. In fact we'll use this stuff for registered tokens that the user didn't interact with.

File: cypress/e2e/00-welcome.cy.js
Comment (raul-oliveira) [reply cypress/e2e/00-welcome.cy.js:1]: It was breaking after some changes, that's why I fixed it here.

File: src/screens/SendTokens.js (hunk 12 above)
Comment (raul-oliveira) [reply src/screens/SendTokens.js:53]: Nice catch. It was an untested scenario.

Now it's working correctly
Image

File: src/components/ModalTransactionOverview.js (hunk 15 above)
Comment (raul-oliveira) [reply src/components/ModalTransactionOverview.js:272]: https://github.com/user-attachments/assets/02170c94-4806-4921-ade2-061fe1b6c5bc

File: src/hooks/useTokenDetails.js (hunk 10 above)
Comment (raul-oliveira) [reply src/hooks/useTokenDetails.js:20]: The hook is fetching from memory store (wallet-lib) token info about everything. The case that we need to fallback is for registered tokens that the user's wallet didn't interact with, so we need the api call.

File: src/components/ModalTransactionOverview.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/ModalTransactionOverview.js:87]: Done @tuliomir 

Image

File: src/components/ModalSendSuccess.js (hunk 13 above)
Comment (raul-oliveira) [reply src/components/ModalSendSuccess.js:82]: Test with Number.MAX_SAFE_INTEGER amount
image

File: src/components/OutputsWrapper.js (hunk 11 above)
Comment (raul-oliveira) [reply src/components/OutputsWrapper.js]: image

File: src/components/SendTokensOne.js [hunk 17]
```diff
@@ -99,6 +124,210 @@ class SendTokensOne extends React.Component {
...
+   * Calculate fee and change output based on input selection mode
+   */
+  calculateFeeAndChange = async () => {
+    const token = this.state.selected;
+    if (!token) return;
+
+    // Token version must always be available
+    if (token.version === undefined) {
+      throw new Error(`Token version not loaded for ${token.uid}. This should never happen.`);
+    }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/SendTokensOne.js:143]: This scenario should never happen, if so we want the error.

File: src/screens/SendTokens.js [hunk 18]
```diff
@@ -105,6 +113,39 @@ function SendTokens() {
...
+
+    return '';
+  };
+
+  /**
+   * Check if user has enough HTR to pay total fee
+   */
+  useEffect(() => {
+    setFeeError(validateFeeBalance());
+  }, [tokenFees, tokensBalance]);
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/screens/SendTokens.js]: done

File: src/components/SendTokensOne.js (hunk 16 above)
Comment (raul-oliveira) [reply src/components/SendTokensOne.js]: done in https://github.com/HathorNetwork/hathor-wallet/pull/858/commits/e827f80ab0b3569644391da52b6feba3d84fe2b4

File: src/components/SendTokensOne.js [hunk 19]
```diff
@@ -99,28 +125,232 @@ class SendTokensOne extends React.Component {
...
+        fee: 0n,
+        isDepositToken: false,
+        changeOutput: null,
+        selectedUtxos: [],
+        error: t`Insufficient balance for ${token.symbol}`
+      });
+      if (this.props.onFeeChange) {
+        this.props.onFeeChange(token.uid, 0n, null);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/SendTokensOne.js:260]: we'll rely in the debouncer for this one.

File: src/screens/SendTokens.js [hunk 20]
```diff
@@ -105,6 +114,39 @@ function SendTokens() {
...
+
+    if (totalFee > 0n) {
+      const htrBalance = tokensBalance[hathorLib.constants.NATIVE_TOKEN_UID]?.data?.available || 0n;
+
+      if (htrBalance < totalFee) {
+        return t`Insufficient HTR balance to pay network fee`;
+      }
+    }
+
+    return '';
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/screens/SendTokens.js]: Done

File: src/screens/SendTokens.js [hunk 21]
```diff
@@ -105,6 +114,53 @@ function SendTokens() {
...
+   * Validate if user has enough HTR to pay total fee.
+   * Returns an error message string, or empty string if balance is sufficient.
+   */
+  const validateFeeBalance = useCallback(() => {
+    const fees = Object.values(tokenFees);
+    if (fees.length === 0) {
+      return '';
+    }
+
+    const totalFee = fees.reduce((a, b) => a + b, 0n);
```
Comment (pedroferreira1) [inline src/screens/SendTokens.js]: The `totalFee` must also include data outputs

Comment (raul-oliveira) [reply src/screens/SendTokens.js]: Fixed. Now we count the data outputs and I also fixed the fee message in the modal

**Review by tuliomir — APPROVED** [review]

File: src/components/SendTokensOne.js [hunk 22]
```diff
@@ -99,28 +130,232 @@ class SendTokensOne extends React.Component {
...
+      const tokenBalance = this.props.tokensBalance[token.uid];
+      let error = null;
+      if (totalAmount > 0n && tokenBalance && tokenBalance.status === TOKEN_DOWNLOAD_STATUS.READY) {
+        const available = tokenBalance.data.available;
+        if (totalAmount > available) {
+          error = `Insufficient ${token.symbol} balance.`;
+        }
+      }
+
+      this.setState({ fee: 0n, isDepositToken, changeOutput: null, selectedUtxos: [], error });
```
[the remark this answers is not in the document]
Comment (raul-oliveira) [reply src/components/SendTokensOne.js:200]: done

**Review by pedroferreira1 — APPROVED** [review]
